msgpack_convert.c

```diff
@@ -471,228 +340,112 @@ int msgpack_convert_object(zval *return_value, zval *tpl, zval **value)
             MSGPACK_WARNING(
                 "[msgpack] (%s) Invocation of %s's constructor failed",
                 __FUNCTION__, ce->name);
 
             return FAILURE;
         }
-        if (retval_ptr)
-        {
-            zval_ptr_dtor(&retval_ptr);
-        }
-        if (params)
-        {
-            efree(params);
-        }
     }
 
-    switch (Z_TYPE_PP(value))
-    {
+    switch (Z_TYPE_P(*value)) {
         case IS_ARRAY:
         {
-            char *key;
-            uint key_len;
-            int key_type;
-            ulong key_index;
-            zval **data;
-            HashPosition pos;
-            HashTable *ht, *ret;
-            HashTable *var = NULL;
+            HashTable *ht, *ret, *var = NULL;
             int num;
+            zend_string *str_key;
+            zval *data;
+            ulong num_key;
 
             ht = HASH_OF(*value);
             ret = HASH_OF(return_value);
 
             num = zend_hash_num_elements(ht);
-            if (num <= 0)
-            {
-                zval_ptr_dtor(value);
+            if (num <= 0) {
+                zval_ptr_dtor(*value);
                 break;
             }
 
             /* string - php_only mode? */
-            if (ht->nNumOfElements != ht->nNextFreeElement
-                || ht->nNumOfElements != ret->nNumOfElements)
-            {
+            if (ht->nNumOfElements != ht->nNextFreeElement || ht->nNumOfElements != ret->nNumOfElements) {
                 HashTable *properties = NULL;
                 HashPosition prop_pos;
 
                 ALLOC_HASHTABLE(var);
                 zend_hash_init(var, num, NULL, NULL, 0);
 
-                zend_hash_internal_pointer_reset_ex(ht, &pos);
-                for (;; zend_hash_move_forward_ex(ht, &pos))
-                {
-                    key_type = zend_hash_get_current_key_ex(
-                        ht, &key, &key_len, &key_index, 0, &pos);
-
-                    if (key_type == HASH_KEY_NON_EXISTANT)
-                    {
-                        break;
-                    }
-
-                    if (zend_hash_get_current_data_ex(
-                            ht, (void *)&data, &pos) != SUCCESS)
-                    {
-                        continue;
-                    }
-
-                    if (key_type == HASH_KEY_IS_STRING)
-                    {
-                        zval *val;
-                        MSGPACK_CONVERT_COPY_ZVAL(val, data);
-                        if (msgpack_convert_string_to_properties(
-                                return_value, key, key_len, val, var) != SUCCESS)
-                        {
-                            zval_ptr_dtor(&val);
-                            MSGPACK_WARNING(
-                                "[msgpack] (%s) "
-                                "illegal offset type, skip this decoding",
-                                __FUNCTION__);
+                ZEND_HASH_FOREACH_STR_KEY_VAL(ht, str_key, data) {
+                    if (str_key) {
```

Comment:
I *think* ZEND_HASH_FOREACH_STR_KEY_VAL will return data that is a numeric key and str_key will just be null.

ZEND_HASH_FOREACH_STR_KEY_VAL just sets up pointers and ZEND_HASH_FOREACH isn't doing any filtering.